stout/task.h

```diff
@@ -72,223 +84,298 @@ struct HeapTask {
+      Interrupt& interrupt,
+      std::exception_ptr&& fail_exception,
+      std::conditional_t<
+          std::is_void_v<Value_>,
+          Callback<>,
+          Callback<Value_>>&& start,
+      Callback<std::exception_ptr>&& fail,
+      Callback<>&& stop) {
+    start_ = std::move(start);
+    fail_ = std::move(fail);
+    stop_ = std::move(stop);
 
-////////////////////////////////////////////////////////////////////////
+    // TODO(benh): clarify the semantics of whether or not calling
+    // 'Register()' more than once is well-defined.
+    adaptor_.Register(interrupt);
 
-template <typename K_, typename Value_, typename... Args_>
-struct _TaskWith {
-  template <typename... Args>
-  void Start(Args&&...) {
-    std::apply(
-        [&](auto&&... args) {
-          start_(
-              std::forward<decltype(args)>(args)...,
-              e_,
-              *interrupt_,
-              [this](auto&&... args) {
-                k_.Start(std::forward<decltype(args)>(args)...);
-              },
-              [this](std::exception_ptr e) {
-                k_.Fail(std::move(e));
-              },
-              [this]() {
-                k_.Stop();
-              });
-        },
-        std::move(args_));
+    adaptor_.Fail(std::move(fail_exception));
   }
 
-  template <typename... Args>
-  void Fail(Args&&... args) {
-    // TODO(benh): propagate through 'Task'.
-    k_.Fail(std::forward<Args>(args)...);
-  }
+  void Stop(
+      Interrupt& interrupt,
+      std::conditional_t<
+          std::is_void_v<Value_>,
+          Callback<>,
+          Callback<Value_>>&& start,
+      Callback<std::exception_ptr>&& fail,
+      Callback<>&& stop) {
+    start_ = std::move(start);
+    fail_ = std::move(fail);
+    stop_ = std::move(stop);
 
-  void Stop() {
-    // TODO(benh): propagate through 'Task'.
-    k_.Stop();
-  }
+    // TODO(benh): clarify the semantics of whether or not calling
+    // 'Register()' more than once is well-defined.
+    adaptor_.Register(interrupt);
 
-  void Register(Interrupt& interrupt) {
-    interrupt_ = &interrupt;
-    k_.Register(interrupt);
+    adaptor_.Stop();
   }
 
-  K_ k_;
-
-  std::tuple<Args_...> args_;
-
-  Callback<
-      Args_&&...,
-      std::unique_ptr<void, Callback<void*>>&,
-      Interrupt&,
-      Callback<Value_>&&,
-      Callback<std::exception_ptr>&&,
-      Callback<>&&>
+  std::conditional_t<
+      std::is_void_v<Value_>,
+      Callback<>,
+      Callback<Value_>>
       start_;
+  Callback<std::exception_ptr> fail_;
+  Callback<> stop_;
 
-  std::unique_ptr<void, Callback<void*>> e_;
+  using Adaptor_ = decltype(std::declval<E_>().template k<void>(
+      std::declval<Adaptor<Value_>>()));
 
-  Interrupt* interrupt_ = nullptr;
+  Adaptor_ adaptor_;
 };
 
 ////////////////////////////////////////////////////////////////////////
 
-} // namespace detail
-
-////////////////////////////////////////////////////////////////////////
+struct _TaskWith {
+  enum class Action {
```

Comment:
Let's add a small comment about why we have this enum and how it works with "dispatch". Thanks!